src/util.cc

```diff
@@ -2414,52 +2446,75 @@ string json_escape_utf8(const string& val)
 	auto val_size = val.length();
 
 	// Reserve at least the size of the existing string to avoid resizing the string in the best-case
-	// scenario where we don't have any multi-byte characters.
-	string result;
-	result.reserve(val_size);
-
-	size_t idx;
-	for ( idx = 0; idx < val_size; )
+	// scenario where we don't have any multi-byte characters. We keep two versions of this string:
+	// one that has a valid utf8 string and one that has a fully-escaped version. The utf8 string gets
+	// returned if all of the characters were valid utf8 sequences, but it will fall back to the
+	// escaped version otherwise. This uses slightly more memory but it avoids looping through all
+	// of the characters a second time in the case of a bad utf8 sequence.
+	string utf_result;
+	utf_result.reserve(val_size);
+	string escaped_result;
+	escaped_result.reserve(val_size);
+
+	bool found_bad = false;
+	size_t idx = 0;
+	while ( idx < val_size )
 		{
 		const char ch = val[idx];
 
 		// Normal ASCII characters plus a few of the control characters can be inserted directly. The
 		// rest of the control characters should be escaped as regular bytes.
-		if ( ( ch >= 32 && ch <= 127 ) ||
+		if ( ( ch >= 32 && ch < 127 ) ||
 		       ch == '\b' || ch == '\f' || ch == '\n' || ch == '\r' || ch == '\t' )
 			{
-			result.push_back(ch);
+			if ( ! found_bad )
+				utf_result.push_back(ch);
+
+			escaped_result.push_back(ch);
 			++idx;
 			continue;
 			}
-		else if ( ch >= 0 && ch < 32 )
+		else if ( found_bad )
 			{
-			result.append(json_escape_byte(ch));
+			// If we already found a bad UTF8 character (see check_ok_utf8) just insert the bytes
+			// as escaped characters into the escaped result and move on.
+			escaped_result.append(json_escape_byte(ch));
 			++idx;
 			continue;
 			}
 
-		// Find out how long the next character should be.
-		unsigned int char_size = getNumBytesForUTF8(ch);
-
-		// If it says that it's a single character or it's not an valid string UTF8 sequence, insert
-		// the one escaped byte into the string, step forward one, and go to the next character.
-		if ( char_size == 0 || idx+char_size > val_size || isLegalUTF8Sequence(val_data+idx, val_data+idx+char_size) == 0 )
+		// If we haven't found a bad UTF-8 character yet, check to see if the next one starts a
+		// UTF-8 character. If not, we'll mark that we're on a bad result. Otherwise we'll go
+		// ahead and insert this character and continue.
+		if ( ! found_bad )
 			{
-			result.append(json_escape_byte(ch));
-			++idx;
-			continue;
-			}
+			// Find out how long the next character should be.
+			unsigned int char_size = getNumBytesForUTF8(ch);
 
-		result.append(val, idx, char_size);
-		idx += char_size;
+			// If we don't have enough data for this character or it's an invalid sequence,
+			// insert the one escaped byte into the string and go to the next character.
+			if ( idx+char_size > val_size ||
```

Comment:
Just double-checking because I can't make up my mind: this shouldn't be `idx+char_size >= val_size`, correct?

Reply:
If val_size is 2 (the string is one two-byte UTF8 sequence long), the index is zero, and the char_size is 2 (for the two bytes in the sequence), we'd want it to be processed as UTF8. If we set that `>=`, it would get escaped incorrectly.